objects/client.c

```diff
@@ -723,12 +723,58 @@
  * identifier remain the same. This allow to match a spawn event to an actual
  * client.
  *
+ * This is used to display a different mouse cursor when the application is
+ * loading and also to attach some properties to the newly created client (like
+ * a `tag` or `floating` state).
+ *
+ * Some applications, like `xterm`, don't support startup notification. While
+ * not perfect, the addition the following code to `rc.lua` will mitigate the
+ * issue. Please note that this code is Linux specific.
+ *
+ *    local blacklisted_snid = setmetatable({}, {__mode = "v" })
+ *
+ *    --- Make startup notification work for some clients like XTerm. This is ugly
+ *    -- but works often enough to be useful.
+ *    local function fix_startup_id(c)
+ *        -- Prevent "broken" sub processes created by `c` to inherit its SNID
+ *        if c.startup_id then
+ *            blacklisted_snid[c.startup_id] = blacklisted_snid[c.startup_id] or c
+ *            return
+ *        end
+ *
+ *        if not c.pid then return end
+ *
+ *        -- Read the process environment variables
+ *        local f = io.open("/proc/"..c.pid.."/environ", "rb")
+ *
+ *        -- It will only work on Linux, that's already 99% of the userbase.
+ *        if not f then return end
+ *
+ *        local snid = f:read("*all"):match("STARTUP_ID=([^\0]*)\0")
```

Comment:
Sigh. Greetings from Lua 5.1:
```
$ lua5.1 -e 'print(io.open("/proc/self/environ"):read("*all"):match("STARTUP_ID=([^\0]*)\0"))' 
lua5.1: (command line):1: malformed pattern (missing ']')
stack traceback:
	[C]: in function 'match'
	(command line):1: in main chunk
	[C]: ?
```

Reply:
Oh. In Lua 5.1, `\z` works for matching a zero byte. However, it does not work in Lua 5.2 or 5.3. (I did not test LuaJIT)

Reply:
So, can I revert to letting `strdup` do its well documented job or should I add code to find the `\0` and cut the string in lua for the sake of pedantism?

Reply:
Pick what you prefer. It feels like all versions are awful :-(.
(You could also do the following, which is still awful)
```lua
local value = _VERSION <= "Lua 5.1" and "([^\z]*)\0" or "([^\0]*)\0"
local snid = f:read("*all"):match("STARTUP_ID=" .. value)
```